docs - models (#19750)
metabase/metabase: docs/users-guide/04-asking-questions.md

- Question
- [SQL query](./writing-sql.md)

This page covers how to ask a question using Metabase's graphical query builder, the "Question" option.

## Creating a new question with the query builder

From the **+ New** dropdown, select **Question**, then pick your starting data:

You can start a question from:

- **A model**. A [model][model] is a special kind of saved question meant to be used as a good starting point for questions. Sometimes these are called derived tables, as they usually pull together data from multiple raw tables.
- **Raw data**. You'll need to specify the database and the table in that database as the starting point for your question.
- A **saved question**. You can use the results of any question as the starting point for a new question.

Note that there are some kinds of saved questions that can't be used as source data:

- Druid questions
- Google Analytics questions
- Mongo questions
- Questions that use `Cumulative Sum` or `Cumulative Count` aggregations
- Questions that have columns that are named the same or similar thing, like `Count` and `Count 2`

## The query builder

> Once you select your data, Metabase will take you to the notebook editor of the query builder. Say you selected **Raw data** > **Sample Dataset** > **Orders**, then you'll see something like this:

Reviewer: > the notebook editor of the query builder

This feels awkward. I wonder if we need the term "notebook editor" anymore, or if it's just "the query builder" now that it's the default pace you're dropped in when doing a GUI query.

![Metabase query builder](./images/asking-questions/notebook-editor.png)

This is the query builder's notebook editor. It has three default steps.

- [Picking data](#picking-data)
- [Filtering](#filtering)
- [Summarizing and grouping by](#summarizing-and-grouping-by)

To the right of completed step is a **Preview** button (looks like a Play button - a triangle pointing to the right) that shows you the first 10 rows of the results of your question up to that step.
 
![Previewing results](./images/notebook/preview-table.png)

## Picking data

> The data section is where you select the data you want to work with. Here you'll pick a [Model][model], a table from a database, or a saved question. You can click on a table to select which columns you want to include in your results.

Reviewer: I don't think models need to be a proper noun.

```suggestion
The data section is where you select the data you want to work with. Here you'll pick a [model][model], a table from a database, or a saved question. You can click on a table to select which columns you want to include in your results.
```

> You can also select multiple tables from the same data source by [joining them](./join.md).

Reviewer: ```suggestion
You can also select multiple tables from the same database by [joining them](./join.md).
```

![Multiple groupings](./images/notebook/histogram-bins.png)

> Once you're done setting your metrics and groupings, click **Done** to close the Summarize sidebar and see your results in all their glory.

Reviewer: This doc feels like it's combining the "simple mode" and notebook experiences now in a couple places in a potentially confusing way. E.g., if you've been following along and are trying to summarize in the notebook, there is no "Summarize sidebar."

## Visualizing your data

If you want to jump ahead and learn about [how to change the visualization](05-visualizing-results.md) of your results, by all means, feel free.

## Returning to the notebook editor

To return to the notebook editor for a question, click on the show editor button in the upper right.

![Show editor](./images/asking-questions/show-editor.png)

## Viewing an individual record's details

To see more info about a given record (a user, order, venue, etc), click on a record's ID number (or primary key). You can see all fields related to that one record and all connected tables that are hidden in the table view for the sake of readability. To page through the other records in the current table, press the right or left arrow keys, or click on the arrows to the right or left of the screen.

![Record details](./images/notebook/record-details.png)

## Custom expressions

![Custom expression](./images/expressions/aggregation-expression.png)

> Custom expressions allow you to use spreadsheet-like functions and simple arithmetic within or between aggregation functions. For example, you could do `Average(sqrt[FieldX]) + Sum([FieldY])` or `Max(floor([FieldX] - [FieldY]))`, where `FieldX` and `FieldY` are fields in the currently selected table. [Learn more about writing expressions](./expressions.md)

Reviewer 2: ```suggestion
Custom expressions allow you to use spreadsheet-like functions and simple arithmetic within or between aggregation functions. For example, you could do `Average(sqrt[FieldX]) + Sum([FieldY])` or `Max(floor([FieldX] - [FieldY]))`, where `FieldX` and `FieldY` are fields in the currently selected table. [Learn more about writing expressions](./expressions.md).
```

### Creating custom columns

![Custom column](./images/expressions/custom-column.png)

Custom columns are helpful when you need to create a new column based on a calculation, such as subtracting the value of one column from another, or extracting a portion of an existing text column. Note that columns you add in a custom question are not permanently added to your table; they'll only be present in the given question.

> You can use the following math operators in your formulas: `+`, `–`, `*` (multiplication), and `/` (division), along with a whole host of spreadsheet-like functions. You can also use parentheses to clarify the order of operations. You can [learn more about writing expressions here](./expressions.md).

Reviewer 2: ```suggestion
You can use the following math operators in your formulas: `+`, `–`, `*` (multiplication), and `/` (division), along with a whole host of spreadsheet-like functions. You can also use parentheses to clarify the order of operations. 
```

Reviewer 2: Seems redundant since this was linked right above; maybe it's useful since they're different sections but as is looks repetitive

## Sorting results

![Sorting](./images/notebook/sort-step.png)

The sorting step lets you pick one or more columns to sort your results by. For each column you pick, you can also choose whether to sort ascending or descending; just click the arrow to change from ascending (up arrow) to descending (down arrow).

## Setting a row limit

The row limit step lets you limit how many rows you want from the previous results. When used in conjunction with sorting, this can let you do things like create a top-10 list, by first sorting by one of the columns in your result, then adding a row limit of 10. Unlike other steps, the row limit step can only be added at the end of your question.

## Viewing the SQL that powers your question

> Under the hood, all Metabase questions are SQL (gasp!). If you're curious to see the SQL that will get run when you ask your question, you can click the little console icon in the top-right of the notebook editor. In the modal that opens up, you'll also be given the option to start a new query in the SQL editor, using this generated SQL as a starting point. It's a nice little shortcut to have Metabase write some boilerplate SQL for you, but then allows you to tweak and customize the query.

Reviewer 2: Is it worth mentioning that depending on permissions, you may not have access to the SQL editor as a user?
